lifetimes/fitters/pareto_nbd_fitter.py

```diff
@@ -282,3 +282,76 @@ def expected_number_of_purchases_up_to_time(self, t):
         first_term = r * beta / alpha / (s - 1)
         second_term = 1 - (beta / (beta + t)) ** (s - 1)
         return first_term * second_term
+
+    def conditional_probability_of_n_purchases_up_to_time(self, n, t, frequency, recency, T):
+        """
+        Return the probability of n purchases up to time t for an individual
+        with history frequency, recency and T (age).
+        From paper:
+        http://www.brucehardie.com/notes/028/pareto_nbd_conditional_pmf.pdf
+        Parameters:
+            n: int
+                number of purchases.
+            t: a scalar
+                time up to which probability should be calculated.
+            frequency: float
+                historical frequency of customer.
+            recency: float
+                historical recency of customer.
+            T: float
+                age of the customer.
+        Returns: a scalar or array
+        """
+        if t <= 0:
+            return 0
+
+        x, t_x = frequency, recency
+        params = self._unload_params('r', 'alpha', 's', 'beta')
+        r, alpha, s, beta = params
+
+        if alpha < beta:
+            min_of_alpha_beta, max_of_alpha_beta, p, p_l_1, p_l_2 = (alpha, beta, r + x + n, r + x, r + x + 1)
+        else:
+            min_of_alpha_beta, max_of_alpha_beta, p, p_l_1, p_l_2 = (beta, alpha, s + 1, s + 1, s)
+        abs_alpha_beta = max_of_alpha_beta - min_of_alpha_beta
+
+        log_l = self._conditional_log_likelihood(params, x, t_x, T)
+        log_p_zero = gammaln(r + x) + r * log(alpha) + s * log(beta) - (
+            gammaln(r) + (r + x) * log(alpha + T) + s * log(beta + T) +
+            log_l
+        )
+        log_B_one = gammaln(r + x + n) + r * log(alpha) + s * log(beta) - (
+            gammaln(r) + (r + x + n) * log(alpha + T + t) + s * log(beta + T + t)
+        )
+        log_B_two = r * log(alpha) + s * log(beta) + gammaln(r + s + x) + betaln(r + x + n, s + 1) + \
+            log(hyp2f1(r + s + x, p, r + s + x + n + 1, abs_alpha_beta / (max_of_alpha_beta + T))) - (
+                gammaln(r) + gammaln(s) + (r + s + x) * log(max_of_alpha_beta + T)
+        )
+
+        def _log_B_three(i):
+            return r * log(alpha) + s * log(beta) + gammaln(r + s + x + i) + betaln(r + x + n, s + 1) + \
+                log(hyp2f1(r + s + x + i, p, r + s + x + n + 1, abs_alpha_beta / (max_of_alpha_beta + T + t))) - (
+                    gammaln(r) + gammaln(s) + (r + s + x + i) * log(max_of_alpha_beta + T + t)
+            )
+
+        zeroth_term = (n == 0) * (1 - exp(log_p_zero))
+        first_term = n * log(t) - gammaln(n + 1) + log_B_one - log_l
```

Comment:
I got rid of your `_log_factorial` function, as `gammaln` is the same thing but faster.